### PR TITLE
o1vm/riscv32im: adding programs for unit tests

### PR DESCRIPTION
**Pull request by dannywillems**
The idea of this set of changes is to introduce all the tooling that can be used by the developers working on the riscv32im to test the interpreter.

The set contains targets in the Makefile to compile the RISC-V GNU tool chain and some examples of basic programs written in the RISC-V 32IM ISA.

The test suite could contain actual functional tests, but also tests that should trigger exceptions or detect malicious prover/witness builders. The idea came out after witnessing these kinds of [subtle changes](https://github.com/o1-labs/proof-systems/pull/2836#discussion_r1866004935).

The test related to [o1vm/resources/programs/riscv32im/src/sll.S](https://github.com/o1-labs/proof-systems/pull/2866/files#diff-4422c205fc930aa143954bf60de0f27e87cc42833c3bf97fc87b8c560aff19f3) is commented for now because `lui` is implemented late in the @svv232 PR train (see https://github.com/o1-labs/proof-systems/pull/2852), and trying to rebase on top of his branch cause conflicts.

I would suggest having this set of changes on master, and start integrating instructions by instructions, with their corresponding unit test? No strong opinion, but it helps to shape the correctness of the slightly bigger program which is fibonnaci.

### CODE REVIEW

**Comment by dannywillems** [comment]
Done.

File: o1vm/README.md [hunk 1]
```diff
@@ -175,3 +175,45 @@ RUN_WITH_CACHED_DATA="y" FILENAME="env-for-latest-l2-block.sh" O1VM_FLAVOR="pick
...
+
+The toolchain will be available in the directory
+`_riscv32-gnu-toolchain/build` at the root of this repository (see variable
+`RISCV32_TOOLCHAIN_PATH` in the [Makefile](../Makefile).
+
+To compile on of the source files available in
+`resources/programs/riscv32im/src`, the user can use:
+
+```shell
+FILENAME=sll.s
```
Comment (martyall) [inline o1vm/README.md:206]:
```suggestion
FILENAME=sll.S
```

currently the extension is capitalized

File: o1vm/README.md [hunk 2]
```diff
@@ -175,3 +175,45 @@ RUN_WITH_CACHED_DATA="y" FILENAME="env-for-latest-l2-block.sh" O1VM_FLAVOR="pick
...
+respectively in `resources/programs/[ISA]/src` and
+`resources/programs/[ISA]/bin`.
+
+### RISC-V 32 bits (riscv32i, riscv32im)
+
+For the RISC-V 32 bits architecture, the user can install the toolchain by using
+`make setup-riscv32-toolchain`.
+
+If you encounter any issue with the build dependencies, you can refer to [this
+GitHub repository](https://github.com/riscv-collab/riscv-gnu-toolchain).
```
Comment (martyall) [inline o1vm/README.md:196]:
```suggestion
GitHub repository](https://github.com/riscv-collab/riscv-gnu-toolchain?tab=readme-ov-file#prerequisites).
```

File: o1vm/README.md [hunk 3]
```diff
@@ -175,3 +175,45 @@ RUN_WITH_CACHED_DATA="y" FILENAME="env-for-latest-l2-block.sh" O1VM_FLAVOR="pick
...
+`_riscv32-gnu-toolchain/build` at the root of this repository (see variable
+`RISCV32_TOOLCHAIN_PATH` in the [Makefile](../Makefile).
+
+To compile on of the source files available in
+`resources/programs/riscv32im/src`, the user can use:
+
+```shell
+FILENAME=sll.s
+
+_riscv32-gnu-toolchain/build/riscv32-unknown-elf-as \
```
Comment (martyall) [inline o1vm/README.md:208]:
```suggestion
_riscv32-gnu-toolchain/build/riscv32-unknown-elf-as/bin/as \
```

This is where the executable installed for me

**Review by martyall — COMMENTED** [review]
Noticed a few things when walking through the o1vm README

File: o1vm/README.md (hunk 3 above)
Comment (dannywillems) [reply o1vm/README.md:208]: Interesting. Which platform?

**Review by martyall — APPROVED** [review]

**Pull request merged**